### DISCUSSION TITLE
Port forwarding

### DISCUSSION THREAD

**Discussion by wlandau** [r-lib/mirai#24] [Q&A] [answered]
Related to https://github.com/shikokuchuo/mirai/discussions/22#discussioncomment-5128933: is there a port forwarding mechanism for `server()` in which a server can use a different port number than the client? Like SSH port forwarding where ports can listen to each other even though the port numbers do not have to agree? That would allow me to use `parallelly::freePort()` on the server as well as the client in order to be 100% sure all ports I attempt to use are free. This could come up in cases I have seen where jobs on a cluster share the same node. Say the client has port 5000 free, but someone else's job on a remote cluster node is already using port 5000, and an error or worse happens because figurative wires get crossed.

**Comment by shikokuchuo** [comment, marked as answer]
NNG always uses an ephemeral port for outbound dialers, this is assigned automatically by the IP stack, so one less thing you need to worry about.

Crossed wires is not possible - as we are using TCP, which is a bytestream - it is for ordered transmission and not lossy like UDP. Further NNG adds much stricter guarantees through its own scalability protocols on top of the transport layer.

On the client, instead of using `parallelly::freePort()`, you can also just specify port 0 on the client e.g. `daemons("tcp://xx.xxx.xxx.175:0")`. To retrieve the actual port used, you can on the socket:
```r
nanonext::parse_url(nanonext::opt(sock[["listener"]][[1]], "url"))[["port"]]
```

**Reply by shikokuchuo** [reply]
In the latest nanonext (0.7.3.9004), I have added a new interface to get the port more directly:
```r
nanonext::opt(sock[["listener"]][[1]], "local-address")[["port"]]
```

**Reply by wlandau** [reply]
What about the array of ports blocked by the client for the active queue? Based on my tests, I think `is_error_value(daemons()$nodes)` is `FALSE` if `mirai` could not reserve the ports, but I can only do that after the sockets are created. I was hoping to use `parallelly::freePort()` to check the whole port range in advance, but I am running into https://github.com/HenrikBengtsson/parallelly/issues/101.

**Reply by shikokuchuo** [reply]
I am not sure I follow. if I use something blatantly wrong such as `daemons("tcp://:22", nodes = 2)`, the queue process will error and exit when attempting to create the socket and listen at the address.
Then if you query it, `is_error_value(daemons()$nodes)` will be TRUE as connections will be 0.

**Reply by wlandau** [reply]
Yes, that's very useful. I was wondering if there is anything in NNG/`nanonext`/`mirai` which will allow users to decide the whole range of ports before calling `daemons("tcp://:22", nodes = 2)` in the first place.

**Reply by wlandau** [reply]
I would like to automate the port selection process in an unobtrusive way. I like how `daemons("tcp://:22", nodes = 2)` works, but because it launches a background process for the active queue, it would be cumbersome to iteratively retry until a random port range is successful.

**Reply by shikokuchuo** [reply]
You're in luck, I've implemented this in 8a31123. Now you can supply a zero port for NNG to auto-assign as per the above e.g. `daemons("tcp://:0")` or `daemons(c("tcp://:0", "tcp://:0") , nodes = 2)` and subsequent calls to `daemons()` will report the actual assigned port number in the URL.

Note for the active queue, you will need to supply a vector of URLs otherwise it will just assign a contiguous block of ports starting from the random one.

**Reply by shikokuchuo** [reply]
In case you need to extract the port numbers from the URL, the URL parsing utility in nanonext should be faster than using a regex `nanonext::parse_url()`.

**Reply by wlandau** [reply]
Wonderful! This completely solves the problem and allows me to go forward with the active queue with more confidence.

**Reply by wlandau** [reply]
By the way, what port range does NNG use for this exactly?

**Reply by shikokuchuo** [reply]
It is only documented that it is the ephemeral range. I didn't find any reference to the exact range - but I did find a new way of getting just the port number!

```r
nanonext::opt(sock[["listener"]][[1]], "tcp-bound-port")
```
